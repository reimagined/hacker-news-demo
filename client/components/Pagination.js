import React from 'react'
import { Link } from 'react-router-dom'
import styled, { css } from 'styled-components'

import Splitter from './Splitter'

const Wrapper = styled.div`
  margin-left: 3em;
  padding: 0.5em 0;
`

const Href = styled.div`
  display: inline;
  font-weight: bold;
  color: #000;
  text-decoration: none;
  &:hover {
    text-decoration: underline;
  }

  ${props =>
    props.disabled &&
    css`
      pointer-events: none;
      cursor: default;
      color: gray;
    `};
`

const Pagination = ({ page, hasNext, location }) => {
  if (page === 1 && !hasNext) {
    return null
  }

  const prevDisabled = page <= 1
  const nextDisabled = !hasNext

  return (
    <Wrapper>
      <Link to={`${location}/${Number(page) - 1}`}>
        <Href disabled={prevDisabled}>Prev</Href>
      </Link>
      <Splitter />
      {page}
      <Splitter />
      <Link to={`${location}/${Number(page) + 1}`}>
        <Href disabled={nextDisabled}>More</Href>
      </Link>
    </Wrapper>
  )
}

Pagination.defaultProps = {
  page: 1
}

export default Pagination
