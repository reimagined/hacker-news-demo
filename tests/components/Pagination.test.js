import React from 'react'
import { shallow } from 'enzyme'

import Pagination from '../../client/components/Pagination'

it('Page 0 renders correctly', () => {
  const markup = shallow(<Pagination page={0} hasNext={true} location="news" />)

  expect(markup).toMatchSnapshot()
})

it('Page 1 renders correctly', () => {
  const markup = shallow(
    <Pagination page={1} hasNext={false} location="news" />
  )

  expect(markup).toMatchSnapshot()
})

it('Page 2 renders correctly', () => {
  const markup = shallow(<Pagination page={2} hasNext={true} location="news" />)

  expect(markup).toMatchSnapshot()
})

it('Page 3 renders correctly', () => {
  const markup = shallow(
    <Pagination page={3} hasNext={false} location="news" />
  )

  expect(markup).toMatchSnapshot()
})

it('Page Default renders correctly', () => {
  const markup = shallow(<Pagination hasNext={false} location="news" />)

  markup.instance().scrollUp()

  expect(markup).toMatchSnapshot()
})
