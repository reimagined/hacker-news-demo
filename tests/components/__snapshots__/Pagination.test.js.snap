// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page 1 renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": null,
  "nodes": Array [
    null,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Pagination
        hasNext={false}
        location="news"
        page={1}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "hasNext": false,
          "location": "news",
          "page": 1,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": null,
        "_debugID": 2,
        "_renderedOutput": null,
      },
      "_renderedNodeType": 2,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Pagination
    hasNext={false}
    location="news"
    page={1}
/>,
}
`;

exports[`Page 1 renders correctly 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <Link
        replace={false}
        to="news/0"
    >
        <styled.div
            disabled={true}
        >
            Prev
        </styled.div>
    </Link>
    <styled.div />
    1
    <styled.div />
    <Link
        replace={false}
        to="news/2"
    >
        <styled.div
            disabled={false}
        >
            More
        </styled.div>
    </Link>
</styled.div>,
  "nodes": Array [
    <styled.div>
      <Link
            replace={false}
            to="news/0"
      >
            <styled.div
                  disabled={true}
            >
                  Prev
            </styled.div>
      </Link>
      <styled.div />
      1
      <styled.div />
      <Link
            replace={false}
            to="news/2"
      >
            <styled.div
                  disabled={false}
            >
                  More
            </styled.div>
      </Link>
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Pagination
        hasNext={true}
        location="news"
        page={1}
/>,
      "_debugID": 9,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "hasNext": true,
          "location": "news",
          "page": 1,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 5,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <Link
                    replace={false}
                    to="news/0"
          >
                    <styled.div
                              disabled={true}
                    >
                              Prev
                    </styled.div>
          </Link>
          <styled.div />
          1
          <styled.div />
          <Link
                    replace={false}
                    to="news/2"
          >
                    <styled.div
                              disabled={false}
                    >
                              More
                    </styled.div>
          </Link>
</styled.div>,
        "_debugID": 10,
        "_renderedOutput": <styled.div>
          <Link
                    replace={false}
                    to="news/0"
          >
                    <styled.div
                              disabled={true}
                    >
                              Prev
                    </styled.div>
          </Link>
          <styled.div />
          1
          <styled.div />
          <Link
                    replace={false}
                    to="news/2"
          >
                    <styled.div
                              disabled={false}
                    >
                              More
                    </styled.div>
          </Link>
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Pagination
    hasNext={true}
    location="news"
    page={1}
/>,
}
`;

exports[`Page 2 renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <Link
        replace={false}
        to="news/1"
    >
        <styled.div
            disabled={false}
        >
            Prev
        </styled.div>
    </Link>
    <styled.div />
    2
    <styled.div />
    <Link
        replace={false}
        to="news/3"
    >
        <styled.div
            disabled={false}
        >
            More
        </styled.div>
    </Link>
</styled.div>,
  "nodes": Array [
    <styled.div>
      <Link
            replace={false}
            to="news/1"
      >
            <styled.div
                  disabled={false}
            >
                  Prev
            </styled.div>
      </Link>
      <styled.div />
      2
      <styled.div />
      <Link
            replace={false}
            to="news/3"
      >
            <styled.div
                  disabled={false}
            >
                  More
            </styled.div>
      </Link>
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Pagination
        hasNext={true}
        location="news"
        page={2}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "hasNext": true,
          "location": "news",
          "page": 2,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <Link
                    replace={false}
                    to="news/1"
          >
                    <styled.div
                              disabled={false}
                    >
                              Prev
                    </styled.div>
          </Link>
          <styled.div />
          2
          <styled.div />
          <Link
                    replace={false}
                    to="news/3"
          >
                    <styled.div
                              disabled={false}
                    >
                              More
                    </styled.div>
          </Link>
</styled.div>,
        "_debugID": 4,
        "_renderedOutput": <styled.div>
          <Link
                    replace={false}
                    to="news/1"
          >
                    <styled.div
                              disabled={false}
                    >
                              Prev
                    </styled.div>
          </Link>
          <styled.div />
          2
          <styled.div />
          <Link
                    replace={false}
                    to="news/3"
          >
                    <styled.div
                              disabled={false}
                    >
                              More
                    </styled.div>
          </Link>
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Pagination
    hasNext={true}
    location="news"
    page={2}
/>,
}
`;

exports[`Page 3 renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <Link
        replace={false}
        to="news/2"
    >
        <styled.div
            disabled={false}
        >
            Prev
        </styled.div>
    </Link>
    <styled.div />
    3
    <styled.div />
    <Link
        replace={false}
        to="news/4"
    >
        <styled.div
            disabled={true}
        >
            More
        </styled.div>
    </Link>
</styled.div>,
  "nodes": Array [
    <styled.div>
      <Link
            replace={false}
            to="news/2"
      >
            <styled.div
                  disabled={false}
            >
                  Prev
            </styled.div>
      </Link>
      <styled.div />
      3
      <styled.div />
      <Link
            replace={false}
            to="news/4"
      >
            <styled.div
                  disabled={true}
            >
                  More
            </styled.div>
      </Link>
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Pagination
        hasNext={false}
        location="news"
        page={3}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "hasNext": false,
          "location": "news",
          "page": 3,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <Link
                    replace={false}
                    to="news/2"
          >
                    <styled.div
                              disabled={false}
                    >
                              Prev
                    </styled.div>
          </Link>
          <styled.div />
          3
          <styled.div />
          <Link
                    replace={false}
                    to="news/4"
          >
                    <styled.div
                              disabled={true}
                    >
                              More
                    </styled.div>
          </Link>
</styled.div>,
        "_debugID": 6,
        "_renderedOutput": <styled.div>
          <Link
                    replace={false}
                    to="news/2"
          >
                    <styled.div
                              disabled={false}
                    >
                              Prev
                    </styled.div>
          </Link>
          <styled.div />
          3
          <styled.div />
          <Link
                    replace={false}
                    to="news/4"
          >
                    <styled.div
                              disabled={true}
                    >
                              More
                    </styled.div>
          </Link>
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Pagination
    hasNext={false}
    location="news"
    page={3}
/>,
}
`;

exports[`Page Default renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": null,
  "nodes": Array [
    null,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Pagination
        hasNext={false}
        location="news"
        page={1}
/>,
      "_debugID": 7,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "hasNext": false,
          "location": "news",
          "page": 1,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 4,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": null,
        "_debugID": 8,
        "_renderedOutput": null,
      },
      "_renderedNodeType": 2,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Pagination
    hasNext={false}
    location="news"
    page={1}
/>,
}
`;
